### PR TITLE
[FEATURE] Binary blob support for OGR provider, processing algorithm to extract binaries

### PR DESCRIPTION
**Pull request by nyalldawson**
Instead of converting binary fields to truncated strings, we instead store their contents as QByteArray values, allowing the original binary content to be retrieved. Update/add support also exists for binary fields (for the OGR provider).

This allows for plugins and scripts to utilise binary fields, such as extracting their contents. This PR takes advantage of this new support and adds a new algorithm for extracting binary blobs from tables. 

This algorithm allows users to extract binary fields to files. The filename is set via a data defined expression, so can be taken from a column value or a more complex expression.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Future TODO: create a new edit widget for binary fields, allowing their contents to be saved to a file/loaded from a file.

File: images/themes/default/mIconFieldBinary.svg [hunk 1]
```diff
@@ -0,0 +1,96 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<svg
+   xmlns:dc="http://purl.org/dc/elements/1.1/"
+   xmlns:cc="http://creativecommons.org/ns#"
+   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
+   xmlns:svg="http://www.w3.org/2000/svg"
+   xmlns="http://www.w3.org/2000/svg"
+   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
+   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
+   height="64"
```
Comment (nirvn) [inline images/themes/default/mIconFieldBinary.svg]: @nyalldawson , should this icon be 16 x 16 pixels in size? Also, don't forget to optimize it.

**Comment by luipir** [comment]
great :)

**Comment by m-kuhn** [comment]
Finally :tada:

File: src/core/qgsogrutils.cpp [hunk 2]
```diff
@@ -221,6 +221,17 @@ QVariant QgsOgrUtils::getOgrFeatureAttribute( OGRFeatureH ogrFet, const QgsField
           value = QDateTime( QDate( year, month, day ), QTime( hour, minute, second ) );
       }
       break;
+
+      case QVariant::ByteArray:
+      {
+        int size = 0;
+        const GByte *b = OGR_F_GetFieldAsBinary( ogrFet, attIndex, &size );
+        QByteArray ba = QByteArray::fromRawData( reinterpret_cast<const char *>( b ), size );
+        ba.detach();
```
Comment (m-kuhn) [inline src/core/qgsogrutils.cpp]: Why the detach?

Comment (m-kuhn) [reply src/core/qgsogrutils.cpp]: http://doc.qt.io/qt-5/qbytearray.html#fromRawData

Comment (nyalldawson) [reply src/core/qgsogrutils.cpp]: I'll add a comment here. It's an odd API decision on the qt side.

File: images/themes/default/mIconFieldBinary.svg (hunk 1 above)
Comment (nyalldawson) [reply images/themes/default/mIconFieldBinary.svg]: Done

**Comment by nyalldawson** [comment]
Now comes complete with:

- support for binary fields in vector file writer
- support for binary fields in memory layers
- support for creating new binary fields for compatible ogr data formats
- a new "binary blob" editor widget, which allows extracting the blob contents to a file on disk, and embedding a file's content to a blob field

I think this is ready for merging if someone wants to review!

**Comment by nirvn** [comment]
@nyalldawson , what about an expression function that could return a base64 string of the blob? That'd allow nice stuff to be done (embedding html img in maptips, layout atlases, etc.)

**Comment by nyalldawson** [comment]
@nirvn that's a good idea! Let's get this one merged first and then we can keep extending in more modular PRs
